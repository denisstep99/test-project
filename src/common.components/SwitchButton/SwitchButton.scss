@import '../../common.styles/proteiStyles';

.switch-button {
  display: inline-block;
  overflow: hidden;
  border-radius: $radiusDefault;
  width: 90px;
  user-select: none;
  border: 1px solid $primary;
  color: $primaryDarker;
  background-color: white;

  &_disabled {
    color: $darkGray;
    border-color: $primaryLighter;
    .switch-button__body::before {
      background-color: $primaryLight;
    }
  }

  .switch-button__body-part-off {
    color: $bgMain;
  }

  &_on {
    .switch-button__body::before {
      left: 43px;
    }

    .switch-button__body-part-on {
      color: $bgMain;
    }

    .switch-button__body-part-off {
      color: $primaryDarker;
    }
  }

  .switch-button__body-part {
    transition-duration: 0.2s;
  }

  &__body {
    cursor: pointer;
    height: 40px;

    display: flex;
    position: relative;
  }

  &__body::before {
    content: "";
    border-radius: $radiusDefault;
    display: block;
    width: 47px;
    height: 42px;

    transition-duration: 0.2s;

    background-color: $primary;

    z-index: 1;

    position: absolute;
    left: -1px;
    top: -1px;
  }

  &__body-part {
    flex-basis: 45px;
    flex-shrink: 0;

    z-index: 2;

    display: flex;
    justify-content: center;
    align-items: center;
  }

  &__checkbox {
    display: none;
  }
}