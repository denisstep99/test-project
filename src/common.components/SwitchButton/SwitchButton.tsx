import './SwitchButton.scss';
import * as React from "react";
import cx from 'classnames';
import {v4 as uuid} from "uuid";
import {ChangeEvent, useState} from "react";

export interface ISwitchButtonProps {
    isSwitchedOn: boolean;

    labelOn?: React.ReactNode;
    labelOff?: React.ReactNode;

    className?: string;
    labelFor?: string;
    disabled?: boolean;

    onChange?(isSwitchedOn: boolean): void;
}

export const SwitchButton = React.memo<ISwitchButtonProps>(({
                                                                           labelOn,
                                                                           labelOff,
                                                                           className,
                                                                           labelFor,
                                                                           disabled,
                                                                           onChange,
                                                                           isSwitchedOn,
                                                                       }) => {
    const [switchButtonId] = useState(labelFor || uuid());
    const switchButtonClass = cx(
        'switch-button',
        {'switch-button_disabled': disabled, 'switch-button_on': isSwitchedOn},
        className
    );

    const changeStateHandler = (event: ChangeEvent<HTMLInputElement>) => {
        onChange && onChange(event.target.checked);
    }

    return (
        <span className={switchButtonClass}>
            <label className="switch-button__body" htmlFor={switchButtonId}>
                <div className="switch-button__body-part switch-button__body-part-off">
                    {labelOff}
                </div>
                <div className="switch-button__body-part switch-button__body-part-on">
                    {labelOn}
                </div>
            </label>
            <input
                className="switch-button__checkbox"
                type="checkbox"
                checked={isSwitchedOn}
                disabled={disabled}
                id={switchButtonId}
                onChange={changeStateHandler}
            />
        </span>
    );
});
