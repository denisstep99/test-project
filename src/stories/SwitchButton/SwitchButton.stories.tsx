import React, {useState} from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';

import {SwitchButton} from "../../common.components/SwitchButton/SwitchButton";
import {DecoratorFunction} from "@storybook/addons";
import {StoryFnReactReturnType} from "@storybook/react/dist/ts3.4/client/preview/types";

const labelDecorator: DecoratorFunction<StoryFnReactReturnType> = (Story) => {
    return (
        <div>
            <label htmlFor="my_id">Связанный лейбл</label>
            <br/>
            <Story/>
        </div>
    )
}

const stateDecorator: DecoratorFunction<StoryFnReactReturnType> = (Story, context) => React.createElement(() => {
    const [isSwitchedOn, setIsSwitchedOn] = useState();

    return (
        <Story args={{...context.args,
            isSwitchedOn,
            onChange: setIsSwitchedOn,
        }}/>
    );
});

export default {
    title: 'Example/SwitchButton',
    component: SwitchButton,
    argTypes: {},
    decorators: [stateDecorator]
} as ComponentMeta<typeof SwitchButton>;

const Template: ComponentStory<typeof SwitchButton> = (args) => <SwitchButton {...args} />;

export const Default = Template.bind({});

export const Labeled = Template.bind({});
Labeled.args = {
    labelOff: "OFF",
    labelOn: "ON"
};

export const Disabled = Template.bind({});
Disabled.args = {
    disabled: true,
    labelOff: "OFF",
    labelOn: "ON"
};

export const ConnectedWithLabel = Template.bind({});
ConnectedWithLabel.args = {
    labelFor: "my_id",
    labelOff: <span>{String.fromCodePoint(0x1F620)}</span>,
    labelOn: <span>{String.fromCodePoint(0x1F609)}</span>
};
ConnectedWithLabel.decorators = [stateDecorator, labelDecorator];
