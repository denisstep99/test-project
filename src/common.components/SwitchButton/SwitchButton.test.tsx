import React from 'react';
import {shallow, ShallowWrapper} from 'enzyme';
import {SwitchButton} from './SwitchButton';

describe('SwitchButton component.', () => {
    let switchButton: ShallowWrapper;

    beforeEach(() => {
        switchButton = shallow(<SwitchButton isSwitchedOn labelFor={'my_label'}/>);
    });

    it('Should render correctly', () => {
        expect(switchButton).toMatchSnapshot();
    });

    it('Should change class on check', () => {
        expect(switchButton.hasClass('switch-button_on')).toBeTruthy();

        switchButton.setProps({isSwitchedOn: false});
        expect(switchButton.hasClass('switch-button_on')).toBeFalsy();
    });

    it("Should display text set by params", () => {
        switchButton.setProps({labelOn: "RU", labelOff: "EN"});
        expect(switchButton.text()).toBe("ENRU");
    })

    it('Should disable on disabled param', () => {
        switchButton.setProps({disabled: true});
        expect(switchButton.hasClass('switch-button_disabled')).toBeTruthy();
    });
});
