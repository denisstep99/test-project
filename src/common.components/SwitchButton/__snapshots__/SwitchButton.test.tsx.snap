// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwitchButton component. Should render correctly 1`] = `
<span
  className="switch-button switch-button_on"
>
  <label
    className="switch-button__body"
    htmlFor="my_label"
  >
    <div
      className="switch-button__body-part switch-button__body-part-off"
    />
    <div
      className="switch-button__body-part switch-button__body-part-on"
    />
  </label>
  <input
    checked={true}
    className="switch-button__checkbox"
    id="my_label"
    onChange={[Function]}
    type="checkbox"
  />
</span>
`;
